stl/inc/xutility

```diff
@@ -5035,46 +5078,71 @@ _NODISCARD constexpr auto lexicographical_compare_three_way(
+    // signed _Elem, signed _Ty
     return SCHAR_MIN <= _Val && _Val <= SCHAR_MAX;
 }
 
 template <class _Ty>
-_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, true_type, false_type, true_type) {
+_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, true_type, false_type, true_type, false_type) {
     // signed _Elem, unsigned _Ty, -1 == static_cast<_Ty>(-1)
     return _Val <= SCHAR_MAX || static_cast<_Ty>(SCHAR_MIN) <= _Val;
 }
 
 template <class _Ty>
-_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, true_type, false_type, false_type) {
+_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, true_type, false_type, false_type, false_type) {
     // signed _Elem, unsigned _Ty, -1 != static_cast<_Ty>(-1)
     return _Val <= SCHAR_MAX;
 }
 
 template <class _Ty>
-_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, false_type, true_type, _Any_tag) {
+_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, false_type, true_type, _Any_tag, false_type) {
     // unsigned _Elem, signed _Ty
     return 0 <= _Val && _Val <= UCHAR_MAX;
 }
 
 template <class _Ty>
-_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, false_type, false_type, _Any_tag) {
+_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, false_type, false_type, _Any_tag, false_type) {
     // unsigned _Elem, unsigned _Ty
     return _Val <= UCHAR_MAX;
 }
 
+template <class _Ty>
+_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, _Any_tag, _Any_tag, _Any_tag, true_type) {
+    // bool _Elem
+    return _Val == true || _Val == false;
+}
+
 template <class _InIt, class _Ty>
-_NODISCARD constexpr bool _Within_limits(_InIt, const _Ty& _Val) { // check whether _Val is within the limits of _Elem
-    using _Elem = remove_pointer_t<_InIt>;
+_NODISCARD constexpr bool _Within_limits(const _InIt&, const _Ty& _Val) {
+    // check whether _Val is within the limits of _Elem
+    using _Elem = _Iter_value_t<_InIt>;
     return _Within_limits(_Val, bool_constant<is_signed_v<_Elem>>{}, bool_constant<is_signed_v<_Ty>>{},
-        bool_constant<-1 == static_cast<_Ty>(-1)>{});
+        bool_constant<-1 == static_cast<_Ty>(-1)>{}, bool_constant<is_same_v<_Elem, bool>>{});
 }
 
 template <class _InIt>
-_NODISCARD constexpr bool _Within_limits(_InIt, const bool&) { // bools are always within the limits of _Elem
+_NODISCARD constexpr bool _Within_limits(const _InIt&, const bool&) { // bools are always within the limits of _Elem
     return true;
 }
 
+#ifdef __cpp_lib_byte
+template <class _InIt>
+_NODISCARD constexpr bool _Within_limits(const _InIt&, const byte&) { // bytes are only comparable with other bytes
+    return true;
+}
+#endif // __cpp_lib_byte
+
+template <class _Iter, class _Ty>
+_INLINE_VAR constexpr bool _Memchr_in_find_is_safe =
+    _Iterator_is_contiguous<_Iter>&&
+        disjunction_v<conjunction<is_integral<_Ty>, _Is_character_or_bool<_Iter_value_t<_Iter>>>

```

Comment:
Could we use `_Is_character_or_byte_or_bool` here instead

Reply:
If we use `_Is_character_or_byte_or_bool` it might allow for `find`ing integers in `std::byte` ranges.

Reply:
Now I am wondering what happens with integers of different sizes.

Shouldnt we always have something like `is_same<_Iter_value_t<_Iter>, _Ty>`

Reply:
Integers of different sizes are handled by the `_Within_limits` function.

Reply:
I need to look at it in Detail, but I do not really like that we do essentially the same thing at two  different places 

Could `_Within_limits` simply reject mixing byte with integrals?

Reply:
It certainly could but I doubt it would be simpler.

Reply:
Although I didn't implement the `memchr` optimization in `find`, I added `_Within_limits` to generalize it, much like how it's being further generalized here, so I can explain its intended design. The intent was for `_Within_limits` to handle runtime conditions - i.e. when searching for a value in a range would compile, and would be eligible for the `memchr` optimization, except that the specific value at runtime could never produce a match. (For example, searching for `300u` in a range of `unsigned char`.) It has both performance and correctness consequences (we can skip the entire range when it won't be found, and `static_cast<unsigned char>(300u)` could produce spurious matches so we must avoid that).

(`_Within_limits` is a variant of C++20 `in_range` although I think the new `bool` behavior makes them different.)

In contrast, the metaprogramming that @AdamBucior is extracting as `_Memchr_in_find_is_safe` determines when the optimization is applicable at compiletime - e.g. searching for `const char*` in a range of `string` is inapplicable. Similarly, searching for `byte` in a range of `unsigned char` is inapplicable - that should be a compiler error (and will be, if we let the classic algorithm attempt to use `==`).

Attempting to get `_Within_limits` to reject `byte` would result in code that compiles and always returns `false`, instead of code that fails to compile. That would be a problem.

While it appears that `_Memchr_in_find_is_safe` and `_Within_limits` have similar logic, they're doing different things that shouldn't be mixed. I think that future simplifications to `_Within_limits` (now that we know how to implement `in_range` conveniently, and have `if constexpr`) would reduce the apparent duplication.